src/lib/libast/sfio/sfsetbuf.c

```diff
@@ -518,7 +519,12 @@ size_t	size;	/* buffer size, -1 for default size */
 	f->flags = (f->flags & ~SF_MALLOC)|sf_malloc;
 
 	if(obuf && obuf != f->data && osize > 0 && (oflags&SF_MALLOC))
-	{	free((Void_t*)obuf);
+	{	
+		/* 
+		Invalid memory free:  https://github.com/att/ast/issues/147
+		Removing it should not cause any memory leaks
```

Comment:
It is far from clear this is true. The conditional above suggests this should only be done when in fact the buffer needs to be freed because we've replaced one buffer with another. The problem may be that `SF_MALLOC` may be set in `oflags` when in fact the buffer was not malloced. I think there is a deeper bug here than just a `free()` that isn't needed. Isn't the problem simply that this line

```
hp->histfp= sfnew(NIL(Sfio_t*),hp->histbuff,HIST_BSIZE,fd,SF_READ|SF_WRITE|SF_APPENDWR|SF_SHARE);
```

was assuming `hp->histbuff` was NULL when in fact it could contain a random value because the allocated memory wasn't zeroed? Doesn't the problem go away with the other half of this change?